ICSharpCode.Decompiler/IL/ControlFlow/SwitchAnalysis.cs

```diff
@@ -230,15 +231,36 @@ bool MatchSwitchVar(ILInstruction inst)
 				return inst.MatchLdLoc(out switchVar);
 		}
 
+		bool MatchSwitchVar(ILInstruction inst, out long sub)
+		{
+			if (inst.MatchBinaryNumericInstruction(BinaryNumericOperator.Sub, out var left, out var right) && right.MatchLdcI(out sub))
+				return MatchSwitchVar(left);
+
+			sub = 0;
+			return MatchSwitchVar(inst);
+		}
+		/// <summary>
+		/// Shifts a LongInterval, treating long.MinValue and long.MaxValue like float.Positive/NegativeInfinity
+		/// </summary>
+		LongInterval ShiftInterval(LongInterval interval, long offset)
+		{
+			return new LongInterval(
+				interval.Start == long.MinValue ? long.MinValue : interval.Start + offset,
+				interval.End == long.MinValue ? long.MinValue : interval.End + offset);
+		}
+
 		/// <summary>
 		/// Analyzes the boolean condition, returning the set of values of the interesting
 		/// variable for which the condition evaluates to true.
 		/// </summary>
 		private bool AnalyzeCondition(ILInstruction condition, out LongSet trueValues)
 		{
-			if (condition is Comp comp && MatchSwitchVar(comp.Left) && comp.Right.MatchLdcI(out long val)) {
+			if (condition is Comp comp && MatchSwitchVar(comp.Left, out var sub) && comp.Right.MatchLdcI(out long val)) {
 				// if (comp(V OP val))
 				trueValues = MakeSetWhereComparisonIsTrue(comp.Kind, val, comp.Sign);
+				if (sub != 0)
+					trueValues = new LongSet(trueValues.Intervals.Select(i => ShiftInterval(i, sub)));
```

Comment:
The `ShiftInterval` logic here doesn't seem quite right to me. Is there any reason you did not use `trueValues.AddOffset(sub)`?
`AddOffset` should be a better match for the wrap-around semantics of `sub`. (see also: `SwitchAnalysis.AnalyzeSwitch()`)

Reply:
Roslyn always uses unsigned casts with the sub, but I did just miss `AddOffset`